src/System.Memory/src/System/Buffers/ReadOnlySequence.cs

```diff
@@ -179,15 +179,10 @@ public ReadOnlySequence<T> Slice(long start, SequencePosition end)
         {
             if (start < 0)
                 ThrowHelper.ThrowStartOrEndArgumentValidationException(start);
+            BoundsCheck(_sequenceStart, end);
             BoundsCheck(end, _sequenceEnd);
 
             SequencePosition begin = Seek(_sequenceStart, end, start);
-            object beginObject = begin.GetObject();
-            object endObject = end.GetObject();
-            if (beginObject != endObject)
-            {
-                CheckEndReachable(beginObject, endObject);
```

Comment:
Why can we get away with not doing this check anymore?

cc @pakrym

Reply:
BoundsCheck makes all work.

Reply:
BoundsCheck does not check the same thing `CheckEndReachable` did, it only looks at RuntimeIndex while `CheckEndReachable` was verifying that segments were actually connected.

Reply:
BoundsChecks is checking for lengths only, but not whether the objects themselves are actually within the sequence. I could be wrong though. @pakrym, what was the intention of these checks.

Reply:
CheckEndReachable  is slow checking so I suggest to remove it or move to debug.assert. Should we add such checking to others methods too? For example: Constructor, GetPosition, Seek?

Reply:
TryGet can check end reachable too.

Reply:
Debug asserts are not visible to consumers they get compiled out in Release mode.
We can discuss removing CheckEndReachable but it shouldn't be done in this PR.

In general, I prefer to see fixes separate from perf improvements so we can weight regressions that are happening.
Currently, it's impossible to tell how much we lost because of bound checking VS gained because we dropped expensive reachability checking.